src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/IPropertyLifetimeManager.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public interface IPropertyLifetimeThingie
+    {
+        void Init(object instance, PropertyLifetimeContext context);
+
+        void Save(object instance, PropertyLifetimeContext context);
+    }
+
+    public struct PropertyLifetimeContext
+    {
+        public PropertyLifetimeContext(ITempDataDictionary tempData, ViewDataDictionary viewData)
+        {
+            TempData = tempData;
+            ViewData = viewData;
+        }
+
+        public ViewDataDictionary ViewData { get; }
+
+        public ITempDataDictionary TempData { get; }
+    }
+
+    internal class PropertyLifetimeManager : IPropertyLifetimeThingie
+    {
+        private Dictionary<Type, IList<LifetimeItem>> _lookup = new Dictionary<Type, IList<LifetimeItem>>();
+
+        public bool Initial { get; set; } = true;
+
+        public void Init(object instance, PropertyLifetimeContext context)
+        {
+            var properties = instance.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance)
+               .Where(property =>
+               {
+                   return
+                       property.GetIndexParameters().Length == 0 &&
+                       property.SetMethod != null &&
+                       (property.IsDefined(typeof(ViewDataAttribute)) || property.IsDefined(typeof(TempDataAttribute)));
+               });
+
+            var items = new List<LifetimeItem>();
+            foreach (var property in properties)
+            {
+                var propertyHelper = new PropertyHelper(property);
+                var source = property.IsDefined(typeof(ViewDataAttribute)) ? LifetimeSource.ViewData : LifetimeSource.TempData;
+                object originalValue;
+                if (source == LifetimeSource.TempData)
+                {
+                    if (Initial)
```

Comment:
Consequence of trying to flow `TempData` between a controller and a view. We have to Peek it to avoid erasing it after it's been read by the first thing (the controller)